### PR TITLE
[PM-9375] Build a debug-for-simulator version of the app for easier automated testing

### PR DESCRIPTION
**Pull request by KatherineInCode**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-9375

## 📔 Objective

In order to run automated tests without locally downloading Xcode and running the tests manually, the QA folks need a debug build that can be downloaded and installed on an arbitrary simulator. This flow produces this build as part of the standard CI flow, along with adding a workflow that can be called separately.

I debated between making this a separate flow (and script) instead of just using the current `build.yml`, but ultimately I felt like the differences between them were too great. There are a lot of steps that don't need to happen for a simulator build, and the `xcodebuild` invocation is sufficiently different that it seemed reasonable to keep it separate. I also thought it was more likely that we would potentially pull this invocation out of the `CI main` workflow and set up an alternate set of triggers for it, depending on QA's needs, and having a separately invocable action makes that easier. However, at some point in the future that calculus might change and it might make more sense to merge this with the `build.yml` workflow.

There are also some minor touch-ups to the `build.sh` script: removing dead code, fixing a capitalization error, and adding more explicit logging.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: Scripts/build-sim.sh
Comment (KatherineInCode) [reply Scripts/build-sim.sh:28]: It doesn't actually matter, it just needs something specified. If you just say `platform=iOS Simulator` it indicates that it can't find a simulator that matches those requirements. But the resulting binary isn't to the best of my knowledge constrained to that particular OS or device.

Would it possibly make sense to just pull it from the `.test-*` files anyways, so we don't accidentally run into issues with that later on?

File: Scripts/build-sim.sh
Comment (KatherineInCode) [reply Scripts/build-sim.sh:1]: Ah, point taken. Though per other conversations, I might actually be going the one-workflow-possibly-also-one-script route which would obviate this a bit.

File: Scripts/build-sim.sh
Comment (fedemkr) [reply Scripts/build-sim.sh:28]: Great idea 💡 It would be awesome to pull the OS name + version from the ` .test-*` files since that's the "main" destination when developing and it will always be updated "automatically"👏

Comment (KatherineInCode) [reply Scripts/build-sim.sh:28]: I think that's what I was originally planning on doing and then forgot partway through working this 😅 Should be pretty easy to pull that logic from the `test.yml` workflow

File: .github/workflows/build.yml
Comment (fedemkr) [inline .github/workflows/build.yml:357]: 🤔 I feel a little bit confusing to have the simulator destination to be passed on Production build variant, even if that's not used inside `build.sh`. Although I understand is done this way not to duplicate this step adding more conditions to the workflow. However, it's like one needs to enter `build.sh` to understand the second parameter is only used in "Simulator" build variant which is not ideal IMO.

**Review by fedemkr — APPROVED** [review]
Approving with a little caveat but I'll leave that up to you to whether change it or not.

File: .github/workflows/build.yml
Comment (KatherineInCode) [reply .github/workflows/build.yml:357]: I don't love it, either. The other option is to have different build scripts for the two flows—simulator build or non-simulator build—but that could potentially run into the bifurcation problem. Neither solution seems particularly great to me, so I defaulted to the "one build script" side.

I suppose a third option would be to make it an optional parameter of the build, but then we'd have the step duplication.

@vvolkgang do you have an opinion on this?

Comment (matt-livefront) [reply .github/workflows/build.yml:357]: I think I've used `'generic/platform=iOS Simulator'` as the destination for a simulator build before and then you don't have to specify a specific device and version.

Comment (KatherineInCode) [reply .github/workflows/build.yml:357]: When I was doing it locally I had to specify the whole thing out, but I don't think I had the "generic" term in there 🤔 I'll double-check and see

File: .github/workflows/build.yml
Comment (vvolkgang) [inline .github/workflows/build.yml:360]: Looking at the file as a whole I think there's two things we could do to spare all of these step checks and duplicates:

1. Preliminary step to set vars for file paths, have scripts check environment variables for simulator specific commands. For the purpose of this PR we can focus on Simulator related changes but this can be applied to all Beta / Prod steps too. Most of them could later be moved to GitHub Environments.
2. I was going to suggest rearranging steps and adding an early exit step but, seems like there's [no support yet ](https://github.com/actions/runner/issues/662) for an early exit with a success state. We can only stop with an error `exit 1` or cancel `gh run cancel ${{ github.run_id }} |  gh run watch ${{ github.run_id }}`. 🤔

**Review by vvolkgang — COMMENTED** [review]
I've not gone through the whole PR yet, sharing some early comments.

File: .github/workflows/CI-main.yml
Comment (KatherineInCode) [reply .github/workflows/CI-main.yml:134]: Ah, good call. Might as well hard code that.

File: .github/workflows/build.yml
Comment (KatherineInCode) [reply .github/workflows/build.yml:14]: Hm. I like that idea, and it makes sense, in case we need to run the Beta app on the simulator. I'll chew on the name for this (Physical Device vs. Simulator), and split it out. That'll impact the conditionals, but not in a bad way I don't think

File: .github/workflows/build.yml
Comment (KatherineInCode) [reply .github/workflows/build.yml:342]: I'll adjust this when I make Simulator vs. Device orthogonal to Variant

File: .github/workflows/build.yml
Comment (KatherineInCode) [reply .github/workflows/build.yml:360]: Hm...I'm not sure I understand how abstracting out the file paths or simulator-specific commands would solve the problem 🤔 it's more that just, if we're building for the simulator we don't need to do a bunch of stuff (but we still could, it would just not have an effect), and the `xcodebuild` invocation is slightly different. But I'm also not sure of the full capabilities of GitHub Environments in that regard.

File: .github/workflows/build.yml
Comment (KatherineInCode) [reply .github/workflows/build.yml:14]: I ultimately decided on "build mode" mostly because that's what I had in the build script already for that

File: .github/workflows/build.yml
Comment (KatherineInCode) [reply .github/workflows/build.yml]: Oh hey, with the generic in there it works, and I can just pull it into the build script and not have to pass it in anymore at all

File: .github/workflows/build.yml
Comment (vvolkgang) [reply .github/workflows/build.yml:21]: Ship it 🚀

File: Scripts/build.sh
Comment (vvolkgang) [inline Scripts/build.sh:3]: 🤔 This probably needs to be updated to remove references to build variants, and add the expected values for Build Mode (device / simulator). E.g.

```
# Builds the Bitwarden iOS app.
# If run in Simulator mode, produces an APP file for the iOS Simulator for ease of automated testing.
# If run in Device mode, produces an IPA file that can be uploaded to Testflight.
#
# Usage:
#
#   $ ./build.sh <build_mode>
#
# Where build_mode is one of:
#   - Device: Build for physical iOS devices
#   - Simulator: Build for iOS Simulator
```

File: .github/workflows/build.yml
Comment (vvolkgang) [reply .github/workflows/build.yml:407]: Made this change, leaving this comment open for reference.

File: Scripts/build.sh
Comment (vvolkgang) [reply Scripts/build.sh:3]: Made this change, leaving this comment open for reference.

File: .github/workflows/build.yml
Comment (vvolkgang) [reply .github/workflows/build.yml:375]: Removed it, easy to add back if we need it later down the line.

File: .github/workflows/build.yml
Comment (vvolkgang) [reply .github/workflows/build.yml:391]: I expected it to be trimmed but who knows, waiting for CI to test. Added it because some I had unrelated artifacts upload issues and it hit me that this could be an issue if we need two builds that only change in flags for some odd reason.

Comment (vvolkgang) [reply .github/workflows/build.yml:391]: That made me realise we also need to add something for build mode.

**Review by vvolkgang — DISMISSED** [review]
👍🏾

**Review by vvolkgang — APPROVED** [review]

**Pull request merged**
